Capture investigation referral data

nathanbegbie opened the pull request.

nathanbegbie committed:
Add behaviour tracking app to nurseconnect

- add redirect view to capture investigation UUID
- add middleware to track user behaviour

nathanbegbie committed:
Remove celery task to create AnalyticsRecord

Request and response objects are not JSON serializable, and a single DB write doesn't  seem worth it to use Celery

nathanbegbie committed:
Get rid of duplicated variable visitor_id => visitor_uuid

rudigiesler commented on analytics/models.py:
```diff
@@ -0,0 +1,14 @@
+    method = models.TextField(null=True)
+    title = models.TextField(null=True)
+    visitor_uuid = models.TextField(null=True)
+    user_profile_uuid = models.TextField(null=True)
```
Should we have `null=True` for `TextField`s? We usually only want `blank=True`, so that we only have one way of representing a null value in the database.

rudigiesler commented on analytics/middleware.py:
```diff
@@ -0,0 +1,29 @@
+
+        visitor_uuid = request.COOKIES.get('visitor_uuid', False)
+        if not visitor_uuid:
+            response.set_cookie('visitor_uuid', str(uuid.uuid4()))
```
Just noting that we can also use the request session to store this UUID, or just use the session ID, instead of dealing with cookies ourselves.

rudigiesler commented on analytics/utils.py:
```diff
@@ -0,0 +1,53 @@
+        exclude = [p for p in settings.ANALYTICS_IGNORE_PATH
+                   if request.path.startswith(p)]
+        return any(exclude)
+    return False
```
This will also include page views for people who didn't go through the analytics endpoint. Just want to make sure that that is the desired functionality.

nathanbegbie replied:
Yep, this was intended

nathanbegbie committed:
Use blank=True instead of null=True for TextFields

nathanbegbie committed:
Use session instead of cookie for tracking visitor UUID

session ID is only recorded after this layer of middleware, hence the need for a new UUID

nathanbegbie committed:
Ignore AJAX requests to /menu/ in analytics

nurseconnect makes AJAX calls to menu on every page load for some reason request.is_ajax fails to pick this up in middleware

nathanbegbie committed:
Ensure blank rather than null values for AnalyticsRecord
